modules/internal/String.chpl

```diff
@@ -17,572 +17,1436 @@
+      if lhs._size <= newLength {
+        const newSize = chpl_mem_goodAllocSize(
+            max(newLength+1,
+                (lhs.len * chpl_stringGrowthFactor):int).safeCast(size_t));
+
+        if lhs.owned {
+          lhs.buff = chpl_mem_realloc(lhs.buff, newSize.safeCast(size_t),
+                                      CHPL_RT_MD_STR_COPY_DATA):bufferType;
+        } else {
+          var newBuff = chpl_mem_alloc(newSize.safeCast(size_t),
+                                       CHPL_RT_MD_STR_COPY_DATA):bufferType;
+          memmove(newBuff, lhs.buff, lhs.len.safeCast(size_t));
+          lhs.buff = newBuff;
+          lhs.owned = true;
+        }
+
+        lhs._size = newSize.safeCast(int);
+      }
+      const rhsRemote = rhs.locale.id != chpl_nodeID;
+      if rhsRemote {
+        chpl_string_comm_get(lhs.buff+lhs.len, rhs.locale.id,
+                              rhs.buff, rhsLen.safeCast(size_t));
+      } else {
+        memmove(lhs.buff+lhs.len, rhs.buff, rhsLen.safeCast(size_t));
+      }
+      lhs.len = newLength;
+      lhs.buff[newLength] = 0;
+    }
   }
 
-  inline proc _cast(type t, x: c_string) where t == string {
-    return toString(x);
+  //
+  // Relational operators
+  // TODO: all relational ops other than == and != are broken for unicode
+  // TODO: It probably will be faster to work on a.locale or b.locale rather
+  //       than here. Especially if a or b is large
+  //
+  // When strings are compared, they compare equal up to the point the
+  // characters in the string differ.   At that point, if the encoding of the
+  // next character is numerically smaller it will compare less, otherwise
+  // greater.  If the encoding of one character uses fewer bits, the smaller
+  // one is zero-extended to match the length of the longer one.
+  // For purposes of comparison, if one compared is shorter than the
+  // other, we fake in a NUL (character code 0x0).  Thus, for example:
+  //  "1000" < "101" < "1010".
+  //
+
+  inline proc _strcmp(a: string, b:string) : int {
+    // Assumes a and b are on same locale and not empty
+    var idx: int = 0;
+    while (idx < a.len && idx < b.len) {
+      if a.buff[idx] != b.buff[idx] then return a.buff[idx]:int - b.buff[idx];
+      idx += 1;
+    }
+    // What if the next character in the longer buffer is 0x0?
+    // I just return +1 or -1 to avoid that question.
+    if (idx < a.len) then return 1;
+    if (idx < b.len) then return -1;
+    return 0;
   }
 
-  inline proc _cast(type t, ref x: c_string_copy) where t == string {
-    return toString(x);
+  proc ==(a: string, b: string) : bool {
+    inline proc doEq(a: string, b:string) {
+      // TODO: is it better to have these outside of the do* fns?
+      //       probably would be 2 extra gets worst case, but avoids doing a
+      //       local copy if b is remote
+      if a.len != b.len then return false;
+      return _strcmp(a, b) == 0;
+    }
+
+    if a.locale.id == b.locale.id {
```

Comment:
I'm pretty sure this should be if a.locale.id == chpl_nodeID && b.locale.id == chpl__nodeID